### ISSUE TITLE
Public resources are denied by the default FileSystemLoader

### ISSUE THREAD

**Issue by bobvandevijver** [liip/LiipImagineBundle#881]
| Q | A
| --- | ---
| Bug Report? | yes
| Feature Request? | yes
| BC Break Report? | yes
| Imagine Bundle Version | 1.7.2

With the release of the 1.7.2, the FileSystemLoader no longer accepts images which are not placed in the "save" data roots. However, this breaks the logic of the public resources in your src directory, if they are symlinked in the web directory. 

This asset symlinking is a commonly used option with Symfony, and therefor this should not be considered unsafe. I know it is possible to add the dataroot of these `src` resources to the configuration of this bundle, but in my opinion, that should not be necessary.

In conclusion: images installed in the `web/bundles` folder which are symlinked to the `src` bundles should not be considered unsafe.

**Comment by robfrawley** [comment]
@bobvandevijver You can either configure your data roots explicitly (yes, this does require you specify all resolved symbolic link paths) or you can switch to the old resolver algorithm.

The default moving forward is to consider symbolic links unsafe. While the usage you are describing is safe, how would you expect we determine that *all* symbolic links are safe? We can't: they could point anywhere on the file system and they bypass the intent of the `data_root` value, which is to constrain your files to the configured director[y|ies].

As for automatically detecting "safe" paths, there is no "one" way to setup a Symfony application, and there is no dependency on a `src` directory, simply a common convention. I don't find it unnecessarily complicated to define the paths *you* consider safe, versus making arbitrary decisions for the user.

```yml
# provide an array of paths to use as data roots
liip_imagine:
  loaders:
    default:
      filesystem:
        data_root:
          - "%kernel.root_dir%/../web"
          - "%kernel.root_dir%/../src/AppBundle/Resources/public"
```
```yml
# use the current locator algorithm (default)
liip_imagine:
  loaders:
    default:
      filesystem:
        locator: filesystem
```
```yml
# use the old locator algorithm (pre 1.7.x)
liip_imagine:
  loaders:
    default:
      filesystem:
        locator: filesystem_insecure
```

**Comment by bobvandevijver** [comment]
I do not want to consider all symbolic links safe, only the symbolic links that point to bundles which have a `Resources/public`-folder, which is, considering the Symfony standard, safe for web usage.

Let's consider the `asset:install` command that is shipped with Symfony. Al it effectively does is making the bundles resources accessible through the web directory, with which you can choose different options. You can make hard copies (which takes time, disk space and might leave deleted resources), or you can create symlinks. And this has nothing to do with your PSR choice: Symfony handles this for you.

Bundle resources are pretty common, and used by multiple bundles. And actually, it is quite easy to add them to the safe path by default: simply use the same logic that the assets install command uses ([source](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Command/AssetsInstallCommand.php#L114)): 
```php
foreach ($this->getContainer()->get('kernel')->getBundles() as $bundle) {
            if (!is_dir($originDir = $bundle->getPath().'/Resources/public')) {
                continue;
            }

            // Add logic to add the absolute path to the data_roots
}
```

So, I am not saying that we should blindly allow everything in the `src` directory: no, most certainly not! However, we should trust the `Resources/public` directories by default, hence the Symfony convention.

**Comment by robfrawley** [comment]
You can submit a PR and see what the other maintainers think of this idea, but again, the point I am trying to make is that the allowed data roots should be constrained by the user, regardless of the paths. 

Most people aren't operating on assets checked into their version control system, as these should be processed already, hence their inclusion in VCS. For those who are, they can easily allow these paths using the current config implementation. No one is saying using symbolic links for assets is a bad idea. (Also, what I was trying to describe via the PSR autoloading is that we can't rely on the `src` path existing, as the application source could be anywhere: we *could* pass the `kernel` service, though).

Perhaps the functionality your describing could be turned on via bundle configuration, but I wouldn't want it to be the default setting. What about supporting regular expressions in the data root paths? This would allow you to use a value like `"%kernel.root_dir%/../*/Resources/public"`. That I'm much more comfortable implementing, and it doesn't add additional container requests.

**Comment by bobvandevijver** [comment]
The advantage of using the `kernel` service over using regular expressions (specifically the example you provide) is that with the `kernel` service you only have the enabled bundles that you add to the root paths, instead of adding all bundles.

For you question on the different `foobar.png` files in different folders: that's why you have the different folders, right? You would either request `folder1/foobar.png` or `folder2/foobar.png`, so it would not matter: it is not only `Resources/public` path thats loaded, you access it through the bundle name. For example, I use the following lines (in the case of a AppBundle), with two different filters: 
```twig
asset('bundles/app/img/foo.png')|imagine_filter('filter_high_res')
asset('bundles/app/img/foo.png')|imagine_filter('filter_low_res')
```

On the processed assets: why? You can easily image a usecase (as above) where you only want one high-resolution image in your resources, with the correct resolutions automatically generated when required. Saves quite some time when you need to update the resource, and in my opinion, that's something this bundle excels in.

Let me say again: I absolute agree with the check on safe paths! But ignoring the paths considered safe by Symfony conventions by default is just wrong. And yes, there are multiple ways of solving them without changing anything in the code of this bundle, but I would definitely add the possibility to auto add these safe paths and I would even enable it by default.

I will take a look into making a PR for this, but that might take a few days.

**Comment by robfrawley** [comment]
The way you implemented this in #883 seems logical so I'm happy to sign off on it.
